packages/libraries/cli/src/base-command.ts

```diff
@@ -304,6 +330,29 @@ class ClientError extends Error {
   }
 }
 
+export type Flags<T extends typeof Command> = Interfaces.InferredFlags<
+  (typeof BaseCommand)['baseFlags'] & T['flags']
+>;
+
+export type Args<T extends typeof Command> = Interfaces.InferredArgs<T['args']>;
+
+type OmitNever<T> = { [K in keyof T as T[K] extends never ? never : K]: T[K] };
+
+type InferConstraintSuccessData<$CommandClass extends typeof Command> =
+  'successDataSchema' extends keyof $CommandClass
+    ? $CommandClass['successDataSchema'] extends SuccessSchema
+      ? z.infer<$CommandClass['successDataSchema']>
+      : 'Error: Missing `static successDataSchema = ...` on your command.'
+    : 'Error: Missing `static successDataSchema = ...` on your command.';
+
+type OptionalSuccessSchema = undefined | SuccessSchema;
+
+type SuccessSchema =
+  | z.SomeZodObject
+  | z.ZodUnion<[z.SomeZodObject, z.SomeZodObject]>
+  | z.ZodUnion<[z.SomeZodObject, z.SomeZodObject, z.SomeZodObject]>
+  | z.ZodUnion<[z.SomeZodObject, z.SomeZodObject, z.SomeZodObject, z.SomeZodObject]>;
+
```

Comment:
In general I try to bias toward a "literate" style of code flow when possible meaning higher level concepts precede lower ones. In practice that looks like references progressing downward which is what I've done here.